### ISSUE TITLE
Instrument name abbreviations

### ISSUE THREAD

**Issue by MarkGotham** [cuthbertLab/music21#633]
Better handling of instrument name abbreviations with a regex that can parse typical structures like '{instrument} in {key}' for transposing instruments (in multiple languages) allowing the inclusion of abbreviations like 'a' for 'alto' as distinct from the 'a' in 'Clarinet in A'.

**Comment by jacobtylerwalls** [comment]
Yep, and the reverse like '{key} {instrument}'. However, I'm not sure that we're prevented from handling 'A' -> `Alto` with the abbreviation list because I'm not sure why this comparison is not using `strip()[0]`. As far as I can tell this will always be True for one-word instrument names:

https://github.com/cuthbertLab/music21/blob/942366e46fadca3a73827a360313ab8e0a3d940b/music21/instrument.py#L2090-L2096

**Comment by jacobtylerwalls** [comment]
If we use `strip[0]` we should be able to put `A` in the abbreviations list without ruining Clarinet in A.

**Comment by jacobtylerwalls** [comment]
Never mind, I now see how multiple word substrings are handled by this condition. My suggestion would degrade them.

**Comment by mscuthbert** [comment]
I'm not sure where we are on this issue -- what is the problem and proposed solution?

**Comment by jacobtylerwalls** [comment]
Problem: the abbreviation "A" for Alto can't be added because "Clarinet in A" will normalize to "Alto", since each word of the input string is tested, and the last usually wins:
```
>>> i = instrument.fromString('Clarinet in A')
>>> i
<music21.instrument.Alto 'Clarinet in A'>
```

Solution: None! 

@MarkGotham's proposal is a new regex for instrument keys; I mentioned it will have to handle being inverted.

My first proposed solution interpreted the code comments to be suggesting there was an attempt to measure the number of letters, the longer results being more likely to be correct (clarinet instead of alto) and ties of equal letters breaking for the later one (trumpet in piccolo trumpet). This was wrong--it was measuring the number of words, to ensure that "Tenor Saxophone" beats "Tenor", which my patch broke.

**Comment by mscuthbert** [comment]
I think that "A." and "A" only means Alto [voice] when it's used alone, and even then it's unclear unless one of S, or T appears as another Part name, since people might name their parts A, B, C, D (but likely not to S)  

We are also expecting that in general we're parsing Part Names, not Part Abbreviations.

In reality, these are all convenience measures; people who really care about proper instrument decoding should encode the instrument.  :-)    I don't plan to commit time to parsing all the different things that "A" could mean, etc. anytime in the near (or medium term) future.  Maybe a student project who wants to be triple-joint with music, linguistics, and CS?
